homeassistant/components/mqtt/__init__.py

```diff
@@ -53,25 +54,62 @@
 ATTR_TOPIC = 'topic'
 ATTR_PAYLOAD = 'payload'
 ATTR_PAYLOAD_TEMPLATE = 'payload_template'
-ATTR_QOS = 'qos'
+ATTR_QOS = CONF_QOS
 ATTR_RETAIN = 'retain'
 
 MAX_RECONNECT_WAIT = 300  # seconds
 
+_VALID_QOS_SCHEMA = vol.All(vol.Coerce(int), vol.In([0, 1, 2]))
+_HBMQTT_CONFIG_SCHEMA = vol.Schema(dict)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_CLIENT_ID): cv.string,
+        vol.Optional(CONF_KEEPALIVE, default=DEFAULT_KEEPALIVE):
+            vol.All(vol.Coerce(int), vol.Range(min=15)),
+        vol.Optional(CONF_BROKER): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT):
+            vol.All(vol.Coerce(int), vol.Range(min=1, max=65535)),
+        vol.Optional(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_CERTIFICATE): vol.IsFile,
+        vol.Optional(CONF_PROTOCOL, default=DEFAULT_PROTOCOL):
+            [PROTOCOL_31, PROTOCOL_311],
```

Comment:
That confuses me all the time. I believe within a Schema a list is handled as all accepted values, but once you wrap it inside a Any or All it doesn't get the special treatment and will be handled as a list. See https://github.com/alecthomas/voluptuous/blob/master/voluptuous.py#L718.